Signal/src/ViewControllers/MediaPageViewController.swift

```diff
@@ -0,0 +1,706 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import UIKit
+
+class MediaPageViewController: UIPageViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate, MediaDetailViewControllerDelegate {
+
+    private struct MediaGalleryItem: Equatable {
+        let message: TSMessage
+        let attachmentStream: TSAttachmentStream
+        let viewController: MediaDetailViewController
+
+        var isVideo: Bool {
+            return attachmentStream.isVideo()
+        }
+
+        var image: UIImage {
+            guard let image = attachmentStream.image() else {
+                owsFail("\(logTag) in \(#function) unexpectedly unagble to build attachment image")
+                return UIImage()
+            }
+
+            return image
+        }
+    }
+
+    private var cachedItems: [MediaGalleryItem] = []
+    private var initialItem: MediaGalleryItem!
+    private var currentItem: MediaGalleryItem! {
+        return cachedItems.first { $0.viewController == viewControllers?.first }
+    }
+
+    private let includeGallery: Bool
+    private let thread: TSThread
+
+    private let mediaGalleryFinder: OWSMediaGalleryFinder
+    private let uiDatabaseConnection: YapDatabaseConnection
+
+    private var mediaMessages: [TSMessage] = []
+
+    convenience init(thread: TSThread, mediaMessage: TSMessage) {
+        self.init(thread: thread, mediaMessage: mediaMessage, includeGallery: true)
+    }
+
+    init(thread: TSThread, mediaMessage: TSMessage, includeGallery: Bool) {
+        self.thread = thread
+        self.uiDatabaseConnection = OWSPrimaryStorage.shared().newDatabaseConnection()
+        self.mediaGalleryFinder = OWSMediaGalleryFinder()
+        self.includeGallery = includeGallery
+
+        let kSpacingBetweenItems: CGFloat = 20
+
+        super.init(transitionStyle: .scroll,
+                   navigationOrientation: .horizontal,
+                   options: [UIPageViewControllerOptionInterPageSpacingKey: kSpacingBetweenItems])
+
+        self.dataSource = self
+        self.delegate = self
+
+        uiDatabaseConnection.beginLongLivedReadTransaction()
+
+        if includeGallery {
+            uiDatabaseConnection.read { transaction in
+                // TODO don't read all media messages in at once. Use Mapping?
```

Comment:
I suspect we'll have to use a mapping before shipping this.

Reply:
Yeah, I wouldn't be totally surprised, though I've done some benchmarking with a few thousand attachments on a fast device, and it's not a problem. I'll definitely do benchmarking on a bigger dataset on a slower device to see what we can get away with.